virttest/utils_libvirt/libvirt_disk.py

```diff
@@ -119,3 +127,299 @@ def create_custom_metadata_disk(disk_path, disk_format,
+                                    disk_target="vda", snapshot_chain_lenghth=4):
+    """
+    Create reuse external snapshots
+
+    :param vm: VM instance
+    :param pre_set_root_dir: preset root directory
+    :param skip_first_one: whether skip first image file
+    :param disk_target: disk target
+    :param snapshot_chain_lenghth : snapshot length
+    :return: absolute root path of backing files and snapshot list
+    """
+    if pre_set_root_dir is None:
+        first_disk_source = get_first_disk_source(vm)
+        root_dir = os.path.dirname(first_disk_source)
+    else:
+        root_dir = pre_set_root_dir
+    meta_options = " --reuse-external --disk-only --no-metadata"
+    # Make four external relative path backing files.
+    relative_sub_folders = [chr(letter) for letter in range(ord('a'), ord('a') + snapshot_chain_lenghth)]
+    backing_file_dict = collections.OrderedDict()
+    snapshot_external_disks = []
+    for index in range(len(relative_sub_folders)):
+        key = relative_sub_folders[index]
+        if index == 0 and skip_first_one:
+            continue
+        else:
+            backing_file_dict[key] = "%s.img" % key
+    for key, value in list(backing_file_dict.items()):
+        backing_file_path = os.path.join(root_dir, key)
+        external_snap_shot = "%s/%s" % (backing_file_path, value)
+        snapshot_external_disks.append(external_snap_shot)
+        options = "%s --diskspec %s,file=%s" % (meta_options, disk_target, external_snap_shot)
+        virsh.snapshot_create_as(vm.name, options,
+                                 ignore_status=False,
+                                 debug=True)
+    logging.debug('reuse external snapshots:%s' % snapshot_external_disks)
+    return root_dir, snapshot_external_disks
+
+
+def make_syslink_path_backing_files(pre_set_root_dir, volume_path_list, origin_image_format="qcow2",
+                                    created_image_format="qcow2", syslink_back_chain_lenghth=4):
+    """
+    Create backing chain files of syslink path for one image
+
+    :param pre_set_root_dir: preset root dir
+    :param volume_path_list: volume path list
+    :param origin_image: origin image
+    :param origin_image_format: original image format
+    :param created_image_format: created image format
+    :param syslink_back_chain_lenghth: syslink back chain length
+    :return: absolute path of top active file and syslink back chain files list
+    """
+    root_dir = pre_set_root_dir
+    syslink_folder_list = [chr(letter) for letter in range(ord('a'), ord('a') + syslink_back_chain_lenghth)]
+    backing_file_dict = collections.OrderedDict()
+    # Make external relative path backing files.
+    for index in range(len(syslink_folder_list)):
+        key = syslink_folder_list[index]
+        folder_path = os.path.join(root_dir, key)
+        cmd = "mkdir -p %s" % folder_path
+        process.run(cmd, shell=True, ignore_status=False)
```

Comment:
As above

Reply:
@kylazhang

not anticipated as you suggest when using os.makedirs(directory, exist_ok=True), see previous paste code